custom/enikshay/tasks.py

```diff
@@ -0,0 +1,267 @@
+
+    def get_property(self, property):
+        """
+        Args:
+            name of the case-property
+
+        Returns:
+            value of the episode case-property named 'property'
+        """
+        return self.episode.dynamic_case_properties().get(property)
+
+    @memoized
+    def get_valid_adherence_cases(self):
+        # Returns list of 'adherence' cases of which 'adherence_value' is one of DOSE_KNOWN_INDICATORS
+        indexed_cases = get_reverse_indexed_cases_es(self.case_updater.domain, self.episode.case_id)
+        return [
+            case
+            for case in indexed_cases
+            if (not case.closed and case.type == CASE_TYPE_ADHERENCE and
+                case.dynamic_case_properties().get('adherence_value') in DOSE_KNOWN_INDICATORS)
+        ]
+
+    def get_latest_adherence_case_for_episode(self):
+        """
+        return open case of type 'adherence' reverse-indexed to episode that
+            has the latest 'adherence_date' property of all
+        """
+        # sometime far back
+        latest_date = pytz.UTC.localize(datetime.datetime(1990, 1, 1))
+        latest_case = None
+        for case in self.get_valid_adherence_cases():
+            adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date'))
+            if adherence_date > latest_date:
+                latest_date = adherence_date
+                latest_case = case
+        return latest_case
+
+    def adherence_cases_between(self, cases, start_date, end_date):
+        """
+        Filter given 'adherence' cases between start_date, end_date
+
+        Args:
+            cases: List of 'adherence' cases
+
+        Returns:
+            List of cases that have 'adherence_date' between start_date and end_date
+
+        """
+        open_pertinent_adherence_cases = [
+            case for case in cases
+            if (
+                start_date.astimezone(pytz.UTC) <=
+                parse_datetime(case.dynamic_case_properties().get('adherence_date')).astimezone(pytz.UTC) <=
+                end_date.astimezone(pytz.UTC))
+        ]
+
+        return open_pertinent_adherence_cases
+
+    def count_doses_taken(self, adherence_cases):
+        """
+        Args:
+            adherence_cases: list of 'adherence_cases' with its 'adherence_value' as one of DOSE_KNOWN_INDICATORS
+
+        Returns:
+            total count of adherence_cases excluding duplicates on a given day. If there are
+            two adherence_cases on one day at different time, it will be counted as one
+        """
+        by_date = defaultdict(list)
+        for case in adherence_cases:
+            adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date')).date()
+            adherence_value = case.dynamic_case_properties().get('adherence_value')
+            if adherence_value in DOSE_TAKEN_INDICATORS:
+                by_date[adherence_date].append(case)
+
+        return len(by_date.keys())
```

Comment:
likewise i think maybe this aggregation could be pushed into ES (with the same caveats/complexity mentioned above).